ksqldb-functional-tests/src/test/resources/query-validation-tests/partition-by.json

```diff
@@ -1100,6 +1172,110 @@
+        {"topic": "OUTPUT", "key": {"ID": null, "NAME": "bob"}, "value": null},
+        {"topic": "OUTPUT", "key": {"ID": null, "NAME": null}, "value": null}
+      ],
+      "post": {
+        "sources": [
+          {"name": "OUTPUT", "type": "stream", "schema": "NAME STRING KEY, ID INT KEY, AGE INT"}
+        ]
+      }
+    },
+    {
+      "name": "multiple columns - select * - some key some value - key first",
+      "statements": [
+        "CREATE STREAM INPUT (NAME STRING KEY, ID INT, AGE INT) with (kafka_topic='input', format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM input PARTITION BY name, id;"
+      ],
+      "inputs": [
+        {"topic": "input", "key": "bob", "value": {"ID": 10, "AGE": 30}},
+        {"topic": "input", "key": "bob", "value": {"ID": null, "AGE": 30}},
+        {"topic": "input", "key": null, "value": {"ID": 10, "AGE": 30}},
+        {"topic": "input", "key": "bob", "value": null},
+        {"topic": "input", "key": null, "value": null}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": {"ID": 10, "NAME": "bob"}, "value": {"AGE": 30}},
+        {"topic": "OUTPUT", "key": {"ID": null, "NAME": "bob"}, "value": {"AGE": 30}},
+        {"topic": "OUTPUT", "key": {"ID": 10, "NAME": null}, "value": {"AGE": 30}},
+        {"topic": "OUTPUT", "key": {"ID": null, "NAME": "bob"}, "value": null},
+        {"topic": "OUTPUT", "key": {"ID": null, "NAME": null}, "value": null}
+      ],
+      "post": {
+        "sources": [
+          {"name": "OUTPUT", "type": "stream", "schema": "NAME STRING KEY, ID INT KEY, AGE INT"}
+        ]
+      }
+    },
+    {
+      "name": "multiple columns - some key some value - properly ordered in selection",
+      "statements": [
+        "CREATE STREAM INPUT (NAME STRING KEY, ID INT, AGE INT) with (kafka_topic='input', format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT name, age, id FROM input PARTITION BY name, id;"
+      ],
+      "inputs": [
+        {"topic": "input", "key": "bob", "value": {"ID": 10, "AGE": 30}},
+        {"topic": "input", "key": "bob", "value": {"ID": null, "AGE": 30}},
+        {"topic": "input", "key": null, "value": {"ID": 10, "AGE": 30}},
+        {"topic": "input", "key": "bob", "value": null},
+        {"topic": "input", "key": null, "value": null}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": {"ID": 10, "NAME": "bob"}, "value": {"AGE": 30}},
+        {"topic": "OUTPUT", "key": {"ID": null, "NAME": "bob"}, "value": {"AGE": 30}},
+        {"topic": "OUTPUT", "key": {"ID": 10, "NAME": null}, "value": {"AGE": 30}},
+        {"topic": "OUTPUT", "key": {"ID": null, "NAME": "bob"}, "value": null},
+        {"topic": "OUTPUT", "key": {"ID": null, "NAME": null}, "value": null}
+      ],
+      "post": {
+        "sources": [
+          {"name": "OUTPUT", "type": "stream", "schema": "NAME STRING KEY, ID INT KEY, AGE INT"}
+        ]
+      }
+    },
```

Comment:
Thanks for the test coverage! This is great. Do we also need to audit the test coverage for GROUP BY?

Reply:
We probably should, though since `GROUP BY` doesn't use the same code we should do that separately (Matthias tested it locally and GROUP BYs have this "rearrange" behavior)